Amendments for INFCOM-4 (#268)
wmo-im/wis2-guide: guide/sections/part2/global-services.adoc

==== 2.7.2 Performance management and monitoring of a Global Service

===== 2.7.2.1 General

> The performance management of designated Global Services is governed by §§2.7–2.9 of the Manual on WIS, Volume II. This section of the Guide provides operational guidance on the service level, the methodology used to assess it, the reporting cadence, and the remediation and restoration practice. Section 2.7.2.9 sets out the per-component operational criteria specific to each category of Global Service. The framework set out in this section and in sections 2.7.2.8 and 2.7.2.9 applies for the purposes of the suspension procedure set out in §2.8 of the Manual and the revocation procedure set out in §2.9 of the Manual; routine incident management for Global Services continues to be coordinated by GISCs under §4.7.4 of the Manual.

Author: ```suggestion
The performance management of designated Global Services is governed by 2.7–2.9 of the Manual on WIS, Volume II. This section of the Guide provides operational guidance on the service level, the methodology used to assess it, the reporting cadence, and the remediation and restoration practice. Section 2.7.2.9 sets out the per-component operational criteria specific to each category of Global Service. The framework set out in this section and in sections 2.7.2.8 and 2.7.2.9 applies for the purposes of the suspension procedure set out in §2.8 of the Manual and the revocation procedure set out in §2.9 of the Manual; routine incident management for Global Services continues to be coordinated by GISCs under §4.7.4 of the Manual.
```

* A Global Monitor:
** Should support a minimum of 50 metrics providers;
** Should support 200 simultaneous "access" to the dashboard;
** Could limit the bandwidth usage of the service to 100 Mb/s.

===== 2.7.2.4 Metrics for Global Services

In the following sections, and for each Global Service, a set of metrics is defined. Each Global Service will provide those metrics. They will then be ingested by the Global Monitor.

===== 2.7.2.5 Service level

The service level is met in a calendar month where the designated Global Service is operational for not less than the proportion of that month set out, for the relevant category of Global Service, in the per-component operational criteria in section 2.7.2.9. The conditions under which a Global Service is considered operational are set out in those criteria, and the proportion of operational time over the month is determined in accordance with section 2.7.2.6.

The service level is the reference for the suspension and revocation procedures set out in 2.8 and 2.9 of the Manual; the conditions under which those procedures are initiated are set out in the Manual.

===== 2.7.2.6 Assessment methodology

> For each calendar month, the proportion of time during which the Global Service was operational is determined. The conditions under which a Global Service is considered operational vary by category and are set out in the per-component operational criteria in section 2.7.2.9. The time during which the Global Service is not operational is determined from the operational metrics published by the Global Service and by the Sensor Centres against those criteria, by aggregating the time intervals during which one or more of the per-component operational criteria are not met. Scheduled maintenance windows notified in advance to the WMO Secretariat are excluded from this time, provided that the total duration of such windows does not exceed four hours per calendar month, unless an exception has been approved by SC-IMT. SC-IMT will also exclude hours where Global Service operation is adversely affected by conditions beyond the control of the operator. The Global Service meets the service level for the calendar month where the proportion of operational time is not less than the value set out in section 2.7.2.5 and 2.7.2.9..

Author: ```suggestion
For each calendar month, the proportion of time during which the Global Service was operational is determined. The conditions under which a Global Service is considered operational vary by category and are set out in the per-component operational criteria in section 2.7.2.9. The time during which the Global Service is not operational is determined from the operational metrics published by the Global Service and by the Sensor Centres against those criteria, by aggregating the time intervals during which one or more of the per-component operational criteria are not met. Scheduled maintenance windows notified in advance to the WMO Secretariat are excluded from this time, provided that the total duration of such windows does not exceed four hours per calendar month, unless an exception has been approved by SC-IMT. SC-IMT will also exclude hours where Global Service operation is adversely affected by conditions beyond the control of the operator. The Global Service meets the service level for the calendar month where the proportion of operational time is not less than the value set out in section 2.7.2.5 and 2.7.2.9.
```

** Additional features, including anti-loop detection, notification message format compliance, validation of the published topic, and metrics provision.

* When connected to the local WIS centre broker, the ``wmo_wis2_gb_connected_flag`` will be set to 1. When the connection to the local WIS centre broker is not possible, the ``wmo_wis2_gb_connected_flag`` will be set to 0.
* When receiving a message from a local WIS centre broker or a Global Service broker, the metric ``wmo_wis2_gb_messages_received_total`` will be increased by 1.
* For data notification messages, a Global Broker will check if a discovery metadata record exists corresponding to the topic on which a message has been published. If there is no corresponding discovery metadata record, the Global Broker will discard non-compliant messages and will raise an alert. The metric ``wmo_wis2_gb_messages_no_metadata_total`` will be increased by 1. The Global Broker should not request information from a Global Discovery Catalogue for each notification message but should keep a cache of all valid topics for every ``centre-id``.

> * A Global Broker will check that the topic on which the message is received is valid. If the topic is invalid, the Global Broker will discard non-compliant messages and will raise an alert. The metric ``wmo_wis2_gb_invalid_topic_total`` will be increased by 1.

Author: ```suggestion
* A Global Broker will check that the topic on which the message is received is valid. If the topic is invalid, the Global Broker will discard non-compliant messages and will raise an alert. The metric ``wmo_wis2_gb_messages_invalid_topic_total`` will be increased by 1.
```

> * A Global Broker will check that the topic on which the message is received is valid. If the topic is invalid, the Global Broker will discard non-compliant messages and will raise an alert. The metric ``wmo_wis2_gb_invalid_topic_total`` will be increased by 1.

Author: ```suggestion
* A Global Broker will check that the topic on which the message is received is valid. If the topic is invalid, the Global Broker will discard non-compliant messages and will raise an alert. The metric ``wmo_wis2_gb_messages_invalid_topic_total`` will be increased by 1.
```

> * A Global Broker will validate notification messages against the standard format (see _Manual on WIS_, Volume II – Appendix E. WIS2 Notification Message), discarding non-compliant messages and raising an alert. The metric ``wmo_wis2_gb_invalid_format_total`` will be increased by 1.

Author: ```suggestion
* A Global Broker will validate notification messages against the standard format (see _Manual on WIS_, Volume II – Appendix E. WIS2 Notification Message), discarding non-compliant messages and raising an alert. The metric ``wmo_wis2_gb_messages_invalid_format_total`` will be increased by 1.
```

> * A Global Broker will validate notification messages against the standard format (see _Manual on WIS_, Volume II – Appendix E. WIS2 Notification Message), discarding non-compliant messages and raising an alert. The metric ``wmo_wis2_gb_invalid_format_total`` will be increased by 1.

Author: ```suggestion
* A Global Broker will validate notification messages against the standard format (see _Manual on WIS_, Volume II – Appendix E. WIS2 Notification Message), discarding non-compliant messages and raising an alert. The metric ``wmo_wis2_gb_messages_invalid_format_total`` will be increased by 1.
```

> * For data notification messages, a Global Broker will check if a discovery metadata record exists corresponding to the topic on which a message has been published. If there is no corresponding discovery metadata record, the Global Broker will discard non-compliant messages and will raise an alert. The metric ``wmo_wis2_gb_messages_no_metadata_total`` will be increased by 1. The Global Broker should not request information from a Global Discovery Catalogue for each notification message but should keep a cache of all valid topics for every ``centre-id``.

Author: ```suggestion
* A Global Broker will check if a discovery metadata record exists corresponding to the topic on which a message has been published. If there is no corresponding discovery metadata record, the Global Broker will discard non-compliant messages and will raise an alert. The metric ``wmo_wis2_gb_messages_no_metadata_total`` will be increased by 1. The Global Broker should not request information from a Global Discovery Catalogue for each notification message but should keep a cache of all valid topics for every ``centre-id``.
```

> * A Global Broker will validate monitoring events against the standard format (see _Manual on WIS_, Volume II – Appendix G. WIS2 Monitoring Events), discarding non-compliant messages and raising an alert. The metric ``wmo_wis2_gb_invalid_format_total`` will be increased by 1.

Author: ```suggestion
* A Global Broker will validate monitoring events against the standard format (see _Manual on WIS_, Volume II – Appendix G. WIS2 Monitoring Events), discarding non-compliant messages and raising an alert. The metric ``wmo_wis2_gb_messages_invalid_format_total`` will be increased by 1.
```

* A Global Monitor:
** Should support a minimum of 50 metrics providers;
** Should support 200 simultaneous "access" to the dashboard;
** Could limit the bandwidth usage of the service to 100 Mb/s.

===== 2.7.2.4 Metrics for Global Services

In the following sections, and for each Global Service, a set of metrics is defined. Each Global Service will provide those metrics. They will then be ingested by the Global Monitor.

===== 2.7.2.5 Service level

The service level is met in a calendar month where the designated Global Service is operational for not less than the proportion of that month set out, for the relevant category of Global Service, in the per-component operational criteria in section 2.7.2.9. The conditions under which a Global Service is considered operational are set out in those criteria, and the proportion of operational time over the month is determined in accordance with section 2.7.2.6.

The service level is the reference for the suspension and revocation procedures set out in 2.8 and 2.9 of the Manual; the conditions under which those procedures are initiated are set out in the Manual.

===== 2.7.2.6 Assessment methodology

> For each calendar month, the proportion of time during which the Global Service was operational is determined. The conditions under which a Global Service is considered operational vary by category and are set out in the per-component operational criteria in section 2.7.2.9. The time during which the Global Service is not operational is determined from the operational metrics published by the Global Service and by the Sensor Centres against those criteria, by aggregating the time intervals during which one or more of the per-component operational criteria are not met. Scheduled maintenance windows notified in advance to the WMO Secretariat are excluded from this time, provided that the total duration of such windows does not exceed four hours per calendar month, unless an exception has been approved by SC-IMT. SC-IMT will also exclude hours where Global Service operation is adversely affected by conditions beyond the control of the operator. The Global Service meets the service level for the calendar month where the proportion of operational time is not less than the value set out in section 2.7.2.5 and 2.7.2.9..

Reviewer: Trailing "." on the paragraph.

> For each incident, the operating Member and the Sensor Centres maintain a record of the start time, the end time, and the affected criteria. These incident records, while not themselves binding, support the determination of operational time and provide operational context for the dashboards and statistical summaries provided by the Global Monitors.

Author: @6a6d74 @efucile incident is not mentioned in previous paragraph

Reviewer: ```suggestion
In the event of an incident affecting the operation of Global Service, the operating Member should maintain a record of the start time, the end time, and affected criteria. These incident records, while not themselves binding, support the determination of operational time and provide operational context for the dashboards and statistical summaries provided by the Global Monitors.
```

Reviewer: I don't think the Sensor Centres are maintaining a "record of [...] incidents". Paragraph changed to give "incident" some context.

Author: ok with Enrico

** Additional features, including anti-loop detection, notification message format compliance, validation of the published topic, and metrics provision.

* When connected to the local WIS centre broker, the ``wmo_wis2_gb_connected_flag`` will be set to 1. When the connection to the local WIS centre broker is not possible, the ``wmo_wis2_gb_connected_flag`` will be set to 0.
* When receiving a message from a local WIS centre broker or a Global Service broker, the metric ``wmo_wis2_gb_messages_received_total`` will be increased by 1.
* For data notification messages, a Global Broker will check if a discovery metadata record exists corresponding to the topic on which a message has been published. If there is no corresponding discovery metadata record, the Global Broker will discard non-compliant messages and will raise an alert. The metric ``wmo_wis2_gb_messages_no_metadata_total`` will be increased by 1. The Global Broker should not request information from a Global Discovery Catalogue for each notification message but should keep a cache of all valid topics for every ``centre-id``.

> * A Global Broker will check that the topic on which the message is received is valid. If the topic is invalid, the Global Broker will discard non-compliant messages and will raise an alert. The metric ``wmo_wis2_gb_invalid_topic_total`` will be increased by 1.

Author: `wmo_wis2_gb_invalid_topic_total` does not match metric CSVs. should it be `wmo_wis2_gb_messages_invalid_topic_total`

Author: @tomkralidis

Reviewer 2: It should be `wmo_wis2_gb_messages_invalid_topic_total`

> * A Global Broker will validate notification messages against the standard format (see _Manual on WIS_, Volume II – Appendix E. WIS2 Notification Message), discarding non-compliant messages and raising an alert. The metric ``wmo_wis2_gb_invalid_format_total`` will be increased by 1.

Author: `wmo_wis2_gb_invalid_format_total` not in metrics GH. 
Should it be `wmo_wis2_gb_messages_invalid_format_total`

Author: @tomkralidis

Reviewer 2: It should be `wmo_wis2_gb_messages_invalid_format_total`
And I don't like much _For data notification messages, a_ We never call the WNM "data notification messages". I'd prefer removing that part.

Author: removed For data notification messages

* A Global Broker will validate monitoring events against the standard format (see _Manual on WIS_, Volume II – Appendix G. WIS2 Monitoring Events), discarding non-compliant messages and raising an alert. The metric ``wmo_wis2_gb_invalid_format_total`` will be increased by 1.
* A Global Broker will republish a message or monitoring event only once. It will record the message identifier (``id``) (as defined in the WIS2 Notification Message) of messages already published and will discard subsequent identical messages (those with the same message ``id``). This is the anti-loop feature of the Global Broker.
* When publishing a message to the local broker, the metric ``wmo_wis2_gb_messages_published_total`` will be increased by 1.

> * In order to improve the service level, a WIS centre may decide to operate two nodes sharing the same ``centre-id``. One will be considered as _primary_ and the other one as _backup_. In this case, when the primary WIS centre broker is reachable (the MQTT connection is up and running), the Global Broker will only process the notification messages from this primary node. When the connection is dropped (the MQTT connection is down) for at least one minute, the Global Broker will start processing the notification messages received from the backup WIS centre broker. When the connection to the primary WIS centre broker is active for at least one minute, the Global Broker will stop processing the messages coming from the backup WIS centre broker and will only process the messages received from the primary WIS centre broker. When the Global Broker is processing the messages from the backup WIS centre broker the metric ``wmo_wis2_gb_connected_backup_flag`` will be set to 1. When the backup WIS centre broker is defined but is not used as the source of notification message, the metric ``wmo_wis2_gb_connected_backup_flag`` will be set to 0.

Author: `wmo_wis2_gb_connected_backup_flag` not in metrics GH

Reviewer 2: Now it is. It was not until very recently.

* Each Global Service should aim to achieve at least 99.5% availability of the service it provides. This is not a contractual target. It should be considered by the entity providing the Global Service as a guideline when designing and operating that service.
* Any Global Service operator anticipating an outage that may exceed 4 hours should notify other WIS2 centres (including other Global Services and the WMO Secretariat) in advance to allow them to prepare for the disruption. A monitor message will be published by the Global Monitor informing about the outage.
* A Global Broker:
** Should support a minimum of 200 WIS2 Nodes or Global Services;
** Should support a minimum of 1 000 subscribers;
** Should support the processing of a minimum of 10 000 messages per second.
* A Global Cache:
** Should support a minimum of 100 GB of data in the cache;
** Should support a minimum of 1 000 simultaneous downloads;

> ** Could limit the number of simultaneous connections from a user (known by its originating source IP) to not less than sixteen;

Author: This is not in the INFCOM Word doc

Author: https://github.com/wmo-im/wis2-guide/commit/0080d081c4902f8d5701e58867b11afe248497eb

** Could limit the bandwidth usage of the service to 1 Gb/s.
* A Global Discovery Catalogue:
** Should support a minimum of 20 000 metadata records;

> ** Should support a minimum of 25 requests per second to the API endpoint.

Author: Lines 64-66 are not in the INFCOM Word doc

Author: not added, just moved.

===== 2.7.2.7 Reporting cadence

Operational metrics for each designated Global Service are published in near-real-time by the Global Service and by the Sensor Centres. The Global Monitors provide dashboards and statistical summaries based on these metrics. The monthly service-level results of each designated Global Service, including performance against the service level over the relevant assessment period, are published openly by the WMO Secretariat in accordance with §2.7.3.1 of the Manual, so that the performance of all designated Global Services is publicly visible, and are also communicated directly to each operating Member in accordance with §2.7.3.3 of the Manual. SC-IMT reviews aggregated performance results at each of its meetings and considers any indications of sustained deterioration in accordance with §2.7.3.4 of the Manual.

===== 2.7.2.8 Remediation and restoration

====== 2.7.2.8.1 Deadline for restoring performance

> Following notification under §2.8.2.1 of the Manual, SC-IMT establishes a deadline under §2.8.2.2 by which the operating Member is expected to re-establish the performance of the Global Service so that it meets the service level. When establishing the deadline, SC-IMT considers the nature of the failure to meet the service level, the operational impact on WIS 2.0 and its users, and the typical time required to address comparable issues. The deadline reflects the operational nature of the Global Service: it may be set in hours where the issue can be addressed by routine operational measures, and in days where more substantive corrective action is required, but it should not normally extend beyond what is needed to address the immediate issue.

Author: ```suggestion
Following notification under §2.8.2.1 of the _Manual on WIS_, Volume II,  SC-IMT establishes a deadline under §2.8.2.2 by which the operating Member is expected to re-establish the performance of the Global Service so that it meets the service level. When establishing the deadline, SC-IMT considers the nature of the failure to meet the service level, the operational impact on WIS 2.0 and its users, and the typical time required to address comparable issues. The deadline reflects the operational nature of the Global Service: it may be set in hours where the issue can be addressed by routine operational measures, and in days where more substantive corrective action is required, but it should not normally extend beyond what is needed to address the immediate issue.
```

Author: updated to match the Word doc

==== 2.7.2.9 Per-component operational criteria
 
The operational criteria specific to each category of designated Global Service are set out below. These criteria define what constitutes operational performance for the purposes of the service level and inform the determination of downtime for each component. The service level stated for each category is the assessable floor used for the performance-management procedures of the Manual; it is distinct from, and not higher than, the general availability guidance in 2.7.2.2. The design-capacity figures for each category remain as set out in 2.7.2.2. The detailed assessment methodology and Key Performance Indicators (KPIs) corresponding to each criterion are set out in Annex B, and the detailed technical checks are maintained by the WMO Secretariat on behalf of SC-IMT and published as a technical reference.
 
For a Global Broker, the service level is not less than 98 per cent of operational time per calendar month.
 
For a Global Cache, the service level is not less than 98 per cent of operational time per calendar month.

==== 2.7.3 Global Broker

> ===== 2.7.3.1 Technical considerations

Author: "and monitoring events" not in Word doc

Author: https://github.com/wmo-im/wis2-guide/pull/267

* When receiving a message from a local WIS centre broker or a Global Service broker, the metric ``wmo_wis2_gb_messages_received_total`` will be increased by 1.
* For data notification messages, a Global Broker will check if a discovery metadata record exists corresponding to the topic on which a message has been published. If there is no corresponding discovery metadata record, the Global Broker will discard non-compliant messages and will raise an alert. The metric ``wmo_wis2_gb_messages_no_metadata_total`` will be increased by 1. The Global Broker should not request information from a Global Discovery Catalogue for each notification message but should keep a cache of all valid topics for every ``centre-id``.
* A Global Broker will check that the topic on which the message is received is valid. If the topic is invalid, the Global Broker will discard non-compliant messages and will raise an alert. The metric ``wmo_wis2_gb_invalid_topic_total`` will be increased by 1.

> 

Author: ```suggestion
```

> * Suggest to remove : During the pre-operational phase (2024), a Global Broker will not discard the message but instead will send a message on the `monitor` topic hierarchy to inform the originating centre and its GISC.~~

Author: need to align branch with Word

Author: remove line 137 from branch

* A Global Broker will validate notification messages against the standard format (see _Manual on WIS_, Volume II – Appendix E. WIS2 Notification Message), discarding non-compliant messages and raising an alert. The metric ``wmo_wis2_gb_invalid_format_total`` will be increased by 1.

> * A Global Broker will validate monitoring events against the standard format (see _Manual on WIS_, Volume II – Appendix G. WIS2 Monitoring Events), discarding non-compliant messages and raising an alert. The metric ``wmo_wis2_gb_invalid_format_total`` will be increased by 1.

Author: this is not in the Word doc

Author: https://github.com/wmo-im/wis2-guide/pull/267

* A Global Broker will update its metrics at least every 60 seconds.
* A Global Broker should not reset its metrics counts.

> * All above-defined metrics will be made available on HTTPS endpoints that the Global Monitor will ingest regularly.

Author: not in Word doc

Author: Just an editorial fix.

* There is no requirement for a Global Cache to provide a browsable interface to the files in its repository in order to allow data consumers to discover what content is available. However, a Global Cache may choose to provide such a capability (for example, implemented as a WAF), along with documentation to inform data consumers of how the capability works.
* The default behaviour for a Global Cache is to cache all data published under the ``origin/a/wis2/{centre-id}/data/core`` topic. A data publisher may indicate that data should not be cached by adding the ``"cache": false`` assertion in the WIS2 Notification Message.
* A Global Cache should make sure that data are downloaded in parallel and that downloads are not blocking each other.

> * A Global Cache may decide not to cache data, for example, if the data are considered too large, or if a WIS2 Node publishes an excessive number of small files. If a Global Cache decides not to cache data, it should behave as though the cache property is set to false and shall send a WME (WIS2 Monitoring Event) message on the ``monitor`` topic hierarchy to inform other WIS centres about it. The Global Cache operator should work with the originating WIS2 Node and its GISC to remedy this issue.

Author: this is not in Word doc

Author: approved by Remy and Jeremy in this PR: https://github.com/wmo-im/wis2-guide/pull/249
